paddle/fluid/operators/activation_op.cc

```diff
@@ -504,6 +504,9 @@ class Relu6OpMaker : public framework::OpProtoAndCheckerMaker {
     AddAttr<float>("threshold",
                    "The threshold value of Relu6. Default is 6.0. ")
         .SetDefault(6.0f);
+    AddAttr<bool>("use_mkldnn",
+                  "(bool, default false) Only used in mkldnn kernel")
+        .SetDefault(false);
```

Comment:
Do we need to add attribute here? A lot of activation op already supports the `use_mkldnn` attribute.

Reply:
Without it it crashes. We run modified mobilenet script:
https://github.com/PaddlePaddle/models/blob/a7fb45f650e0d32ba697b5818aa4799c2ccc9b05/dygraph/mobilenet/mobilenet_v2.py#L69
passing use_mkldnn=True in that line.
The error is:
```
Traceback (most recent call last):
  File "train.py", line 262, in <module>
    train_mobilenet()
  File "train.py", line 195, in train_mobilenet
    avg_loss.backward()
  File "</nfs/pdx/home/sfraczek/.local/lib/python2.7/site-packages/decorator.pyc:decorator-gen-97>", line 2, in backward
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/wrapped_decorator.py", line 25, in __impl__
    return wrapped_func(*args, **kwargs)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/framework.py", line 219, in __impl__
    return func(*args, **kwargs)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/dygraph/varbase_patch_methods.py", line 172, in backward
    framework._dygraph_tracer(), retain_graph)
KeyboardInterrupt

Reservation status error
sfraczek@aipg-ra-clx-86:/data/sfraczek/models/dygraph/mobilenet$ bash run_cpu_v2.sh | grep relu
Traceback (most recent call last):
  File "train.py", line 262, in <module>
    train_mobilenet()
  File "train.py", line 180, in train_mobilenet
    out = net(img)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/dygraph/layers.py", line 600, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/data/sfraczek/models/dygraph/mobilenet/mobilenet_v2.py", line 212, in forward
    y = self._conv1(inputs, if_act=True)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/dygraph/layers.py", line 600, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/data/sfraczek/models/dygraph/mobilenet/mobilenet_v2.py", line 69, in forward
    y = fluid.layers.relu6(y, use_mkldnn=True)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/layers/nn.py", line 9359, in relu6
    'use_mkldnn': use_mkldnn})
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/layer_helper.py", line 43, in append_op
    return self.main_program.current_block().append_op(*args, **kwargs)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/framework.py", line 2789, in append_op
    kwargs.get("stop_gradient", False))
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/dygraph/tracer.py", line 45, in trace_op
    not stop_gradient)
paddle.fluid.core_avx.EnforceNotMet:

--------------------------------------------
C++ Call Stacks (More useful to developers):
--------------------------------------------
0   std::string paddle::platform::GetTraceBackString<std::string >(std::string&&, char const*, int)
1   paddle::platform::EnforceNotMet::EnforceNotMet(paddle::platform::ErrorSummary const&, char const*, int)
2   std::conditional<std::is_pointer<paddle::framework::Attribute >::value, bool const*, bool const&>::type paddle::platform::details::SafeBoostGetConst<bool, paddle::framework::Attribute >(paddle::framework::Attribute const&, char const*, char const*, int)
3   paddle::operators::GetKernelType(paddle::framework::ExecutionContext const&, paddle::framework::OperatorWithKernel const&, std::string const&)
4   paddle::operators::ActivationOp::GetExpectedKernelType(paddle::framework::ExecutionContext const&) const
5   paddle::imperative::PreparedOp paddle::imperative::PrepareOpImpl<paddle::imperative::VarBase>(paddle::imperative::details::NameVarMapTrait<paddle::imperative::VarBase>::Type const&, paddle::imperative::details::NameVarMapTrait<paddle::imperative::VarBase>::Type const&, paddle::framework::OperatorWithKernel const&, paddle::platform::Place, paddle::framework::AttributeMap const&)
6   paddle::imperative::PreparedOp::Prepare(paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::OperatorWithKernel const&, paddle::platform::Place const&, paddle::framework::AttributeMap const&)
7   paddle::imperative::Tracer::TraceOp(std::string const&, paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::AttributeMap, paddle::platform::Place const&, bool)

----------------------
Error Message Summary:
----------------------
InvalidArgumentError: boost::get failed, cannot get value (GetAttr(name)) by type bool, its type is int. at (/data/sfraczek/Paddle/paddle/fluid/framework/operator.h:259)
  [operator < relu6 > error]

Reservation status error
sfraczek@aipg-ra-clx-86:/data/sfraczek/models/dygraph/mobilenet$ bash run_cpu_v2.sh | grep relu
Traceback (most recent call last):
  File "train.py", line 262, in <module>
    train_mobilenet()
  File "train.py", line 180, in train_mobilenet
    out = net(img)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/dygraph/layers.py", line 600, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/data/sfraczek/models/dygraph/mobilenet/mobilenet_v2.py", line 212, in forward
    y = self._conv1(inputs, if_act=True)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/dygraph/layers.py", line 600, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/data/sfraczek/models/dygraph/mobilenet/mobilenet_v2.py", line 69, in forward
    y = fluid.layers.relu6(y, use_mkldnn=True)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/layers/nn.py", line 9359, in relu6
    'use_mkldnn': use_mkldnn})
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/layer_helper.py", line 43, in append_op
    return self.main_program.current_block().append_op(*args, **kwargs)
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/framework.py", line 2789, in append_op
    kwargs.get("stop_gradient", False))
  File "/data/sfraczek/Paddle/build/python/paddle/fluid/dygraph/tracer.py", line 45, in trace_op
    not stop_gradient)
paddle.fluid.core_avx.EnforceNotMet:

--------------------------------------------
C++ Call Stacks (More useful to developers):
--------------------------------------------
0   std::string paddle::platform::GetTraceBackString<std::string >(std::string&&, char const*, int)
1   paddle::platform::EnforceNotMet::EnforceNotMet(paddle::platform::ErrorSummary const&, char const*, int)
2   std::conditional<std::is_pointer<paddle::framework::Attribute >::value, bool const*, bool const&>::type paddle::platform::details::SafeBoostGetConst<bool, paddle::framework::Attribute >(paddle::framework::Attribute const&, char const*, char const*, int)
3   paddle::operators::GetKernelType(paddle::framework::ExecutionContext const&, paddle::framework::OperatorWithKernel const&, std::string const&)
4   paddle::operators::ActivationOp::GetExpectedKernelType(paddle::framework::ExecutionContext const&) const
5   paddle::imperative::PreparedOp paddle::imperative::PrepareOpImpl<paddle::imperative::VarBase>(paddle::imperative::details::NameVarMapTrait<paddle::imperative::VarBase>::Type const&, paddle::imperative::details::NameVarMapTrait<paddle::imperative::VarBase>::Type const&, paddle::framework::OperatorWithKernel const&, paddle::platform::Place, paddle::framework::AttributeMap const&)
6   paddle::imperative::PreparedOp::Prepare(paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::OperatorWithKernel const&, paddle::platform::Place const&, paddle::framework::AttributeMap const&)
7   paddle::imperative::Tracer::TraceOp(std::string const&, paddle::imperative::NameVarBaseMap const&, paddle::imperative::NameVarBaseMap const&, paddle::framework::AttributeMap, paddle::platform::Place const&, bool)

----------------------
Error Message Summary:
----------------------
InvalidArgumentError: boost::get failed, cannot get value (GetAttr(name)) by type bool, its type is int. at (/data/sfraczek/Paddle/paddle/fluid/framework/operator.h:259)
  [operator < relu6 > error]

```

Reply:
Got it. I think adding `use_mkldnn` attribute in `activation_op.cc` is OK.